(Replaced) p-ata

rustopian opened the pull request:
(REPLACED due to history complexity and upstream/dep changes)

This is a draft pending testing of the latest optimizations, especially custom entrypoint.

p-ata (pinocchio-ata) is a drop-in replacement for SPL ATA. Following in the footsteps of [p-token](https://github.com/solana-program/token/tree/main/p-token), it uses pinocchio instead of solana-program to reduce compute usage. Plus, it includes a number of additional improvements.

See [README](https://github.com/rustopian/associated-token-account/blob/p-ata/p-ata/README.md) for
* descriptions of new features and improvements
* instructions for running the various test suites
* benchmark numbers
* byte-for-byte SPL ATA vs. p-ATA comparisons

### Optimal bump circumstances *as of 2025-08-11, ecde2cad*

| Test                  | SPL ATA | p-ata | bump arg | all optimizations |  notes                       |
|-----------------------|--------:|------:|---------:|------------------:|-----------------:|
| create_idempotent     |   3669  |  1732 |     590  |               590 |  |
| create                |  12364  |  4842 |    3323  |              3216 |  |
| create_token2022      |  14692  |  7642 |    6123  |              5987 |  |
| create_topup          |  15817  |  4710 |    3191  |              3083 | `CreateAccountPrefunded` |
| create_topup_nocap    |  15817  |  7476 |    5959  |              5851 | no `CreateAccountPrefunded` |
| create_extended       |  17620  |  9771 |    8251  |              7997 |  |
| recover_nested        |  12851  |  8100 |    N/A   |              8100 | rare instruction  |
| recover_multisig      |    N/A  |  8412 |    N/A   |              8412 | rare instruction  |
| worst_case_create     |  19864  | 15187 |    3323  |              3216 | hard-to-find bump |

### Average of 10,000 random wallets *as of 2025-08-11, ecde2cad*

| Test                  | SPL ATA | p-ata | bump arg | all optimizations |  notes                       |
|-----------------------|--------:|------:|---------:|------------------:|-----------------:|
| create_idempotent     |  4914   |  3234 |      948 |               948 |  |
| create                | 14194   |  6343 |     3692 |              3589 |  |
| create_token2022      | 16057   |  9140 |     6491 |              6362 |  |
| create_topup          | 17317   |  6183 |     3561 |              3449 | `CreateAccountPrefunded` |
| create_topup_no_cap   | 17287   |  8987 |     6324 |              6209 | no `CreateAccountPrefunded` |
| create_extended       | 19420   | 11289 |     8618 |              8366 |  |
| recover_nested        | 17066   | 12576 |      N/A |             12576 | rare instruction  |
| recover_multisig      |   N/A   | 12875 |      N/A |             12875 | rare instruction  |

rustopian committed:
adjust comment/table

rustopian committed:
rm some checks

rustopian committed:
workflow handling

rustopian committed:
rm bad failure test

rustopian committed:
start caching work

rustopian committed:
Merge pull request #5 from rustopian/p-ata-dev

rustopian committed:
bench to readme instead of PR comment

rustopian committed:
rm archive

rustopian committed:
submodules recursive in CI

rustopian committed:
reliably build-programs

rustopian committed:
optimal bump so we're not benching varying entropy

rustopian committed:
Merge branch 'p-ata' into p-ata-dev

rustopian committed:
ax double CI

rustopian committed:
Merge branch 'p-ata-dev' of https://github.com/rustopian/associated-token-account into p-ata-dev

rustopian committed:
Merge pull request #6 from rustopian/p-ata-dev

bench to readme and other various updates

rustopian committed:
discm constants

rustopian committed:
better failure evaluation

rustopian committed:
rm comment

rustopian committed:
reduce cheese

rustopian committed:
fmt

rustopian committed:
fix multisig test setup

rustopian committed:
Merge pull request #7 from rustopian/p-ata-dev

discm constants

rustopian committed:
Merge branch 'solana-program:main' into p-ata

rustopian committed:
refactor test pks etc

rustopian committed:
Merge pull request #8 from rustopian/p-ata-dev

refactor test pks etc

rustopian committed:
cleanup

rustopian committed:
Merge pull request #9 from rustopian/p-ata-dev

cleanup

rustopian committed:
saturating sub, rm more pre-refactor code

rustopian committed:
Merge pull request #10 from rustopian/p-ata-dev

saturating sub, rm more pre-refactor code

rustopian committed:
rm old  code

rustopian committed:
clean up failure tests, ata impl code

rustopian committed:
show details when not byte-identical

rustopian committed:
Merge pull request #11 from rustopian/p-ata-dev

Various test fixes

rustopian committed:
cleanup program id and impl stuff

rustopian committed:
Merge pull request #12 from rustopian/p-ata-dev

cleanup program id and impl stuff

rustopian committed:
avoid accidental system program pk in validation test

rustopian committed:
Merge pull request #13 from rustopian/p-ata-dev

fix: avoid accidental system program pk in validation test

rustopian committed:
constants

rustopian committed:
account builders

rustopian committed:
Merge pull request #14 from rustopian/p-ata-dev

account builders for tests

rustopian committed:
account data comparison service

rustopian committed:
common optimal bump across multiple program ids

rustopian committed:
always lowest

rustopian committed:
fix multisig bump test

rustopian committed:
Merge pull request #15 from rustopian/p-ata-dev

More test cleanup, optimal keys across implementations, byte-for-byte

rustopian committed:
improve output

rustopian committed:
improve output summaries

rustopian committed:
Merge pull request #16 from rustopian/p-ata-dev

improved output & summaries

rustopian committed:
failure mode helpers

rustopian committed:
Merge pull request #17 from rustopian/p-ata-dev

failure mode helpers

rustopian committed:
dedup

rustopian committed:
AccountType instead of string, readme

rustopian committed:
bench commit

rustopian committed:
Merge pull request #18 from rustopian/p-ata-dev

dedup, AccountType, README

rustopian committed:
Mention full debug logs in README.md

rustopian committed:
rm some old comments

rustopian committed:
Merge pull request #19 from rustopian/p-ata-dev

rm some old comments

rustopian committed:
fmt, explanatory comment

rustopian committed:
Merge branch 'p-ata' of https://github.com/rustopian/associated-token-account into p-ata

rustopian committed:
fmt, explanatory comment

rustopian committed:
add'l cleanup

rustopian committed:
trim failure builders a bit

rustopian committed:
rm orphan comment

rustopian committed:
extract formatter

rustopian committed:
rm misleading 'adapt_instruction_data'

rustopian committed:
rm misleading 'adapt_instruction_data'

rustopian committed:
refactor mint data, etc.

rustopian committed:
rm comments

rustopian committed:
rm formatter dups, cleanup

rustopian committed:
update benches

rustopian committed:
cargo

rustopian committed:
DRY for failure builders

rustopian committed:
similar helpers for recovernested failures

rustopian committed:
comfy-table just for benching

rustopian committed:
use strum to rm manual name fns

rustopian committed:
versions

rustopian committed:
restore optimal bump for recover tests

rustopian committed:
restore optimal bump for recover tests

rustopian commented:
@febo I'm glad to have `derive_address`, but when `bump` is provided this program doesn't even bother validating with `create_program_address` - just counts on downstream failure later in execution (hence all the failure equivalency tests). Any reason you find this concerning?

https://github.com/rustopian/associated-token-account/blob/8f7c84a4a0acbdedb5de9b34c368945fba3d1c3b/p-ata/src/processor.rs#L186-L195

rustopian committed:
save 6 CUs with unsafe in parse_ata_accounts and 1 in process_instruction

rustopian committed:
fmt clippy

febo commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,451 @@
+/// Build Transfer instruction data
+#[inline(always)]
+fn build_transfer_data(amount: u64) -> [u8; 9] {
+    let mut data = [0u8; 9];
```
nit: same here about zero-initialization.

febo commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,451 @@
+#[inline(always)]
+fn resolve_rent(rent_info_opt: Option<&AccountInfo>) -> Result<Rent, ProgramError> {
+    match rent_info_opt {
+        Some(rent_acc) => unsafe { Rent::from_account_info_unchecked(rent_acc) }.cloned(),
```
nit: it will be better to avoid the `cloned()` when you have the account info.

rustopian committed:
save 60 CUs on create_token2022 with pointer eq

rustopian committed:
unchecked accounts for recover

rustopian commented on p-ata/src/entrypoint.rs:
```diff
@@ -0,0 +1,47 @@
+nostd_panic_handler!();
+
+#[inline(always)]
+pub fn entry(program_id: &Pubkey, accounts: &[AccountInfo], data: &[u8]) -> ProgramResult {
```
Updated in 2e49943

rustopian committed:
no panic in account parse

rustopian commented:
Pulling back to draft as `bump` is being reconsidered; clients feeding in valid but suboptimal bumps would find non-canonical ATA accounts. I'm benchmarking the various ways to keep some of the cost benefits without violating canonicity.

rustopian committed:
need len check for rent, so free good error

rustopian committed:
fmt

rustopian committed:
gt, not eq

rustopian committed:
safety comment

rustopian committed:
nvm, save 2 CUs

joncinque reviewed (commented):
This is definitely on the right track! A few things to fixup, but the overall work is good

rustopian commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,464 @@
+        None => {
+            let (_, computed_bump) = derive_ata_pda(
+                wallet.key(),
+                token_prog.key(),
```
Thank you, had not thought of this corner case! Will fix up and comment.

rustopian commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,464 @@
+            // Calculate correct space: 165 for base TokenAccount, +5 for ImmutableOwner extension
+            if is_token_2022_program(token_prog.key()) {
+                TokenAccount::LEN + 5 // 170 bytes total for Token-2022 with ImmutableOwner
+            } else {
+                TokenAccount::LEN // 165 bytes for regular Token
```
Indeed, to minimize CUs I was requiring length be passed in if it was different. A bit tyrannical to be sure! I gather based on this and other notes that we overall want things more ergonomic by default than that. Will revisit tomorrow.

rustopian commented on p-ata/src/lib.rs:
```diff
@@ -0,0 +1,13 @@
+#![no_std]
+
+extern crate pinocchio;
+extern crate pinocchio_system;
+extern crate pinocchio_token;
+extern crate spl_token_interface;
```
Since Rust 2018, not required. Old habits. Some toolchains still haven't updated 😅

rustopian commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,464 @@
+
+        // Count how many of the provided signer accounts are both marked as
+        // signer on this instruction *and* appear in the multisig signer list.
+        let mut signer_count: u8 = 0;
+        'outer: for signer_acc in signer_infos {
+            if !signer_acc.is_signer() {
+                continue;
+            }
+            for ms_pk in multisig_state.signers[..multisig_state.n as usize].iter() {
+                if ms_pk == signer_acc.key() {
+                    signer_count = signer_count.saturating_add(1);
+
+                    if signer_count >= multisig_state.m {
+                        break 'outer;
+                    }
+                    continue 'outer;
+                }
+            }
+        }
+
+        if signer_count < multisig_state.m {
```
Of course, dup account attack 🙏 fix incoming

rustopian replied:
Added a "must fail" test for this & patched the problem in 0a5364b

rustopian committed:
len match fvalidation in parse_ata_accounts

rustopian committed:
account structs

rustopian committed:
rm dinosaurs

rustopian committed:
clean up not(CreateAccountPrefunded) area

rustopian commented on p-ata/src/tools/account.rs:
```diff
@@ -0,0 +1,104 @@
+                    owner: target_program_owner,
+                }
+                .invoke_signed(&[signer.clone()])?;
+            }
```
Of course. This whole section only exists to bench/test without CreatePrefundedAccount. Cleaned up in d6ec342

rustopian commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,451 @@
+    &'a AccountInfo,
+    &'a AccountInfo,
+    Option<&'a AccountInfo>,
+);
```
Made this and recover accounts both structs in d946d5a

rustopian committed:
multisig dup signer attack should-fail test

rustopian committed:
multisig dup signer attack should-fail test

rustopian committed:
longform var names

rustopian committed:
is_owned_by, wrong multisig wallet owner fail test

rustopian committed:
fail test drains lamprots from uninitialized ata

rustopian committed:
fail test drains lamports from uninitialized ata

rustopian committed:
fail test drains lamprots from uninitialized ata

rustopian committed:
Merge branch 'p-ata' of https://github.com/rustopian/associated-token-account into p-ata

rustopian committed:
ensure payer is signer for create

rustopian committed:
derive_address for create. Baseline of some benching experiments

rustopian committed:
all validation in RecoverNested

rustopian commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,464 @@
+        None => {
+            let (_, computed_bump) = derive_ata_pda(
+                wallet.key(),
+                token_prog.key(),
```
Verifying nested, owner, and destination ATA addresses in `RecoverNested`, with option to pass in `bump` for this instruction removed. Also verifying destination ATA in `Create`, though checking `payer.is_signer()` may be enough.

Once I'm done with all tests that prove attacks fail, will move to optimizing.

rustopian committed:
mark get_token_account_unchecked unsafe

rustopian committed:
rm some debug msgs

rustopian committed:
rm empty comment

rustopian committed:
old tests with mollusk adapter

rustopian committed:
use actual SPL test files with adapter

rustopian committed:
fmt

rustopian committed:
restore build-programs stuff in script

rustopian committed:
rm benchmarks.yml

rustopian committed:
rm old build script

rustopian committed:
rm unused

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,451 @@
+fn resolve_rent(rent_info_opt: Option<&AccountInfo>) -> Result<Rent, ProgramError> {
+    match rent_info_opt {
+        Some(rent_acc) => unsafe { Rent::from_account_info_unchecked(rent_acc) }.cloned(),
+        None => Rent::get(),
```
https://github.com/anza-xyz/solana-sdk/issues/235

rustopian commented on p-ata/src/tests/mollusk_adapter.rs:
```diff
@@ -0,0 +1,556 @@
+    std::vec,
+};
+
+fn process_instruction(
```
Absolutely. Call me paranoid, but I first wanted to quickly ensure that p-ata passes SPL ATA tests without any modifications.

rustopian commented on p-ata/src/entrypoint.rs:
```diff
@@ -1,46 +1,71 @@
 
 #[inline(always)]
-pub fn entry(program_id: &Pubkey, accounts: &[AccountInfo], data: &[u8]) -> ProgramResult {
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &[u8],
+) -> ProgramResult {
+    process_instruction_inner(program_id, accounts, data)
+}
+
+#[inline(always)]
+fn process_instruction_inner(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &[u8],
+) -> ProgramResult {
```
Nope, debugging artifact scheduled for the chopping block for sure.

rustopian commented on p-ata/src/entrypoint.rs:
For sure. I moved the benching to "average x number of runs" to get good numbers with real bump distributions. Collecting updated numbers now

rustopian commented on p-ata/src/processor.rs:
```diff
@@ -1,60 +1,105 @@
+    // SAFETY: Safe because we are comparing the pointers of the
+    // program_id and TOKEN_2022_PROGRAM_ID, which are both const Pubkeys
+    unsafe {
+        core::ptr::eq(
+            program_id.as_ref().as_ptr(),
+            TOKEN_2022_PROGRAM_ID.as_ref().as_ptr(),
```
Not a silly question at all - still an artifact. I'll be more careful about pushing intermediate WIP.

rustopian committed:
start transfer_unchecked, some small cleanup

rustopian committed:
transfer_unchecked, validate_point, and some other cleanup

rustopian committed:
fmt, clippy

rustopian committed:
fix account type

rustopian committed:
rm bump/len from adapter (bigger token accounts supported in next commit)

rustopian committed:
rm old branch

rustopian committed:
initial token_account_len rework work

rustopian committed:
init bump/canonical test

rustopian committed:
fix syscall return parsing

rustopian committed:
only 1 means off-curve

rustopian committed:
update some dep'd imports

rustopian committed:
add immutableownerextensions to token2022 account size cpi

rustopian committed:
check for spl token, not token2022

rustopian committed:
streamline get_token_account_size

rustopian committed:
pub(crate) for create_pda_account

rustopian committed:
fix comment

rustopian committed:
fmt

rustopian committed:
rm old comment

rustopian committed:
if account len is multisig, validate is multisig

rustopian committed:
canonical, not canoncial

rustopian committed:
unit tests for al
l helpers

rustopian committed:
avoid zero init with MaybeUninit

rustopian committed:
avoid unnecessary clone on one branch

rustopian committed:
rm unneeded AccountMeta

rustopian committed:
MAX_SANE_ACCOUNT_LENGTH and tests

rustopian committed:
rm adapter from non_canonical_bump test

rustopian committed:
rm a few old pseudocode comments

rustopian committed:
Rm old test pseudocode

rustopian committed:
rm comments

rustopian committed:
rm comment

rustopian committed:
benches use optimal bump when BENCH_ITERATIONS=1

rustopian committed:
recover -2 CUs with const instruction data

rustopian committed:
in-line token extension lengths

rustopian committed:
inline extension lengths, plus cross-test with token-2022

rustopian committed:
rm unnecessary is_valid_multisig_data

rustopian committed:
only attempt local extension length lookup if token-2022

rustopian committed:
simpler but not more expensive program id comp

rustopian committed:
benches for create_extended with multiple extensions

rustopian committed:
init migrated unit tests

rustopian committed:
modularize unit test utils

rustopian committed:
support all extensions

rustopian committed:
exhaustive extension combo test, dep resolutions

rustopian committed:
exhaustive extension test

rustopian committed:
rm old pseudocomments and extra debug prints for exhaustive test

rustopian committed:
various small test updates

rustopian committed:
Update README.md

rustopian committed:
some optimizing, readme

rustopian committed:
readme benches

rustopian committed:
Header

rustopian committed:
add avg 10000 to readme

rustopian committed:
Update README.md

rustopian committed:
enum for readability, support upcoming extension

rustopian committed:
better name for calculate_account_size_from_mint_extensions

rustopian committed:
Clean up README.md

rustopian committed:
dedup to common util

rustopian committed:
Merge branch 'p-ata' of https://github.com/rustopian/associated-token-account into p-ata

rustopian committed:
odds and ends

rustopian committed:
safer AccountInfo build in parsing tests

rustopian committed:
test_extension_utils.rs

rustopian committed:
test context manager for account infos, etc.

rustopian committed:
update to latest pinocchio, rename to create-prefunded-account

rustopian committed:
factor out size.rs and recover.rs

rustopian committed:
cheaper derivation for CreateIdempotent

rustopian committed:
on-curve create idemp attack test

rustopian committed:
recover_nested bump safety tests

rustopian committed:
doc comments and some var names

rustopian committed:
comments

rustopian committed:
test structure and readme

rustopian committed:
update fn name

rustopian committed:
update benches in README

rustopian committed:
consistency

rustopian committed:
comment on alternative to find_program_address

rustopian committed:
rm verbiage

rustopian committed:
start unification

rustopian committed:
re-export for bench+test workaround

rustopian committed:
ignore blocks for doctests

rustopian committed:
mollusk setup helper

rustopian committed:
use mollusk helper more

rustopian committed:
checkpoint

rustopian committed:
check in

rustopian committed:
check in more unification

rustopian committed:
unify benches into tests fully

rustopian committed:
features std in bench commands for now

rustopian committed:
Merge pull request #20 from rustopian/unify-bencher-with-tests

Unify bencher with tests

rustopian committed:
pk_array macro

rustopian committed:
fix debug logs etc.

rustopian committed:
misc cleanup

rustopian committed:
reduce adapter

rustopian committed:
various reductions

rustopian committed:
Merge pull request #21 from rustopian/reduce-bench-loc

Reduce bench loc

rustopian committed:
all constructed failure scenarios first assert baseline succeeds, before mutating

rustopian committed:
Failure scenarios all ensure baseline succeeds before mutating to failure inputs

Failure scenarios all ensure baseline succeeds before mutating to failure inputs

rustopian committed:
readme

rustopian committed:
build.rs doc comment

joncinque commented on p-ata/src/entrypoint.rs:
```diff
@@ -0,0 +1,126 @@
+/// [3] mint                                        - token mint account
+/// [4] system_program                              - system program
+/// [5] token_program                               - token program
+/// [6] rent_sysvar              (optional)         - rent sysvar (for optimization)
```
This was normally only used by the "deprecated" instruction that provides no data, but we can add it back in if it's an optimization!

joncinque commented on p-ata/src/entrypoint.rs:
```diff
@@ -0,0 +1,126 @@
+                2 => {
+                    return match instruction_data {
+                        [] => process_recover_nested(program_id, accounts, None),
+                        [owner_bump, nested_bump, destination_bump] => process_recover_nested(
```
Hoo boy, I would really prefer to not have these inputs if I'm being honest 😅 this instruction in general should really not be used often, so let's avoid doing anything more than incidental optimizations from switching to pinocchio and all that

joncinque commented on p-ata/src/entrypoint.rs:
```diff
@@ -0,0 +1,126 @@
+                ),
+                // Bump + account_len provided (for Token-2022 optimization)
+                [bump, account_len_bytes @ ..] => {
+                    // SAFETY: runtime-bounded, and account_len is last.
+                    let account_len = unsafe {
+                        u16::from_le_bytes(*(account_len_bytes.as_ptr() as *const [u8; 2]))
+                    };
```
This needs to check that there's at least two bytes, right? We shouldn't read past the provided instruction data

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+#[cfg(target_os = "solana")]
+use pinocchio::syscalls::sol_curve_validate_point;
+
+pub const INITIALIZE_ACCOUNT_3_DISCM: u8 = 18;
+pub const INITIALIZE_IMMUTABLE_OWNER_DISCM: u8 = 22;
+pub const TRANSFER_CHECKED_DISCM: u8 = 12;
```
nit: same here, can spell out `DISCRIMINANT` / `DISCRIMINATOR`

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+///
+/// Safety: caller must ensure account_data.len() >= 109.
+#[inline(always)]
+pub(crate) unsafe fn is_initialized_account(account_data: &[u8]) -> bool {
```
I don't think this is unsafe -- the slice will do the bounds check and panic appropriately. Did you mean to use `get_unchecked`?

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+    false
+}
+
+/// Get zero-copy mint reference from account info
+#[inline(always)]
+pub(crate) unsafe fn get_mint_unchecked(account: &AccountInfo) -> &Mint {
+    let mint_data_slice = account.borrow_data_unchecked();
+    &*(mint_data_slice.as_ptr() as *const Mint)
+}
```
Be careful, this is very unsafe! Can you put a safety warning saying to at least check that it's a mint or that it has 82 bytes?

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+                if !is_off_curve(&maybe_canonical_address)
+                    || maybe_canonical_address != *associated_token_account.key()
+                {
+                    msg!("PROGRAM: addresses don't match");
```
nit: any reason for the `PROGRAM:` prefix in the log message? If not, best to remove, or even better, give more info with those bytes! ie `Provided bump {bump} does not result in canonical or provided address`

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+/// - **Tests**: Uses curve25519-dalek to replicate on-chain behavior  
+/// - **Other builds**: Returns `false`
+#[inline(always)]
+pub(crate) fn is_off_curve(_address: &Pubkey) -> bool {
```
nit: can remove the leading underscore

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+
+        use curve25519_dalek::edwards::CompressedEdwardsY;
+
+        let mut bytes = MaybeUninit::<[u8; 32]>::uninit();
+        let bytes_ptr = bytes.as_mut_ptr() as *mut u8;
+        // SAFETY: We initialize all 32 bytes before calling assume_init()
+        let bytes = unsafe {
+            core::ptr::copy_nonoverlapping(_address.as_ref().as_ptr(), bytes_ptr, 32);
+            bytes.assume_init()
+        };
+        let compressed = CompressedEdwardsY(bytes);
```
Is this necessary? A `clone()` would probably be simpler since `Pubkey` in pinocchio is already an array

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+
+        match compressed.decompress() {
+            None => true,                    // invalid encoding → off-curve
+            Some(pt) => pt.is_small_order(), // small-order = off-curve, otherwise on-curve
```
This might be my ignorance, but where is the `is_small_order` check coming from? We typically just check to see if it decompresses. Here's the syscall code: https://github.com/anza-xyz/agave/blob/d30889f6ddd58d68c3a5448fcd1aebaa96789a99/curves/curve25519/src/edwards.rs#L82

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+/// that only the highest valid bump can be used, maintaining deterministic
+/// address derivation across all clients.
+#[inline(always)]
+pub(crate) fn ensure_no_better_canonical_address_and_bump(
```
nit: this function confused me a bit. It could be the `ensure` part of the function name confused me, or how it was used... how about calling it `try_find_better_program_derived_address_and_bump` and returning `Option<(Pubkey, u8)>`? Then it'll be clear if something better was found

joncinque commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,595 @@
+    if verified_associated_token_account_to_create
+        .is_some_and(|address| &address != create_accounts.associated_token_account_to_create.key())
+    {
+        msg!("Error: Canonical address does not match provided address. Use correct owner and optimal bump.");
```
nit: May as well print that optimal bump!

joncinque reviewed (commented):
Took another look through the program implementation, minus recover_nested -- will do that on the next pass.

Things are looking good!

rustopian committed:
rm unneeded (and wrong) MINT_WITH_TYPE_SIZE

rustopian committed:
rm bumps for RecoverNested

rustopian committed:
update benches after RecoverNested deoptimization

rustopian committed:
various nits, pinocchio_log

rustopian committed:
clean up is_off_curve when not on solana

rustopian committed:
rename to try_find_better_program_derived_address_and_bump, move log

rustopian committed:
rename to reject_if_better_valid_bump_exists

rustopian committed:
get_mint, not get_mint_unchecked

rustopian committed:
get_mint to get_decimals_from_mint

rustopian committed:
unsafe/safe fixes

rustopian committed:
comment on unsafe in size.rs

rustopian committed:
extension constants instead of huge enum

rustopian committed:
idiomatic test/bench tree (#23)

Moved benches to "/benches" and integration tests to "/tests". Unit tests and some common helpers (used by all three) remain in "/src". Benches do not run under #[cfg(test)], so they require the feature std in order to access test helpers.

rustopian committed:
add some more logs before errors

rustopian committed:
safety comment, rm conversion

rustopian committed:
comment on mint data verification not needed, inline

rustopian committed:
recover seeds, bench numbers

rustopian committed:
10000 benches numbers

rustopian committed:
unnecessary clone

febo commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,982 @@
+#[inline(always)]
+pub(crate) fn get_decimals_from_mint(account: &AccountInfo) -> Result<u8, ProgramError> {
+    let mint_data_slice = account.try_borrow_data()?;
+    if mint_data_slice.len() < MINT_BASE_SIZE {
+        log!(
+            "Error: Mint account data too small. Expected at least {} bytes, found {} bytes",
+            MINT_BASE_SIZE,
+            mint_data_slice.len()
+        );
+        return Err(ProgramError::InvalidAccountData);
+    }
+    // SAFETY: We've validated the account length above
+    let mint = unsafe { &*(mint_data_slice.as_ptr() as *const Mint) };
```
Probably can be replaced by:
```suggestion
    let mint = unsafe { spl_token_interface::state::load_unchecked::<Mint>(&mint_data_slice) };
```

febo commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,982 @@
+
+                if canonical_address != *associated_token_account.key() {
+                    log!(
+                        "Error: Provided associated token account address is not canonical. Expected: {}, Found: {}",
```
Same here about the "non-canonical" message – I think this is a mismatch instead of non-canonical.

febo commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,982 @@
+
+    // Initialize ImmutableOwner for non-SPL Token programs (future compatible)
+    if !is_spl_token_program(token_program.key()) {
+        let initialize_immutable_owner_metas = &[AccountMeta {
+            pubkey: associated_token_account.key(),
+            is_writable: true,
+            is_signer: false,
+        }];
```
`AccountMeta` has some helper constructors:
```suggestion
        let initialize_immutable_owner_metas =
            &[AccountMeta::writable(associated_token_account.key())];
```

rustopian committed:
rm needless ok

Co-authored-by: Fernando Otero <febo@anza.xyz>

rustopian committed:
use `AccountMeta` helpers

Co-authored-by: Fernando Otero <febo@anza.xyz>

rustopian committed:
use `AccountMeta` helper

Co-authored-by: Fernando Otero <febo@anza.xyz>

rustopian committed:
use pinocchio seeds helper

Co-authored-by: Fernando Otero <febo@anza.xyz>

rustopian committed:
rm needless ok

Co-authored-by: Fernando Otero <febo@anza.xyz>

rustopian committed:
nits and rework of load_token_account fns

rustopian committed:
use load_unchecked for mint

Co-authored-by: Fernando Otero <febo@anza.xyz>

rustopian committed:
fix missing ?

rustopian committed:
log for mint account data too small

rustopian committed:
Simplify rent in parse_create_accounts

Co-authored-by: Fernando Otero <febo@anza.xyz>

rustopian committed:
fmt

rustopian committed:
improve error logs

febo commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,974 @@
+/// Get token account reference with validation. Fails if a mutable borrow
+/// of the account has occurred.
+#[inline(always)]
+pub(crate) fn load_token_account(account: &AccountInfo) -> Result<&TokenAccount, ProgramError> {
+    let account_data = account.try_borrow_data()?;
+    if !valid_token_account_data(&account_data) {
+        return Err(ProgramError::InvalidAccountData);
+    }
+    // SAFETY: We've validated the account data structure above
+    unsafe { Ok(&*(account_data.as_ptr() as *const TokenAccount)) }
+}
```
Getting the raw pointer (line 164) from the `Ref` (line 159) is problematic. The `Ref` keeps track of borrows, but when this method finished, it will be drop and reset the borrow state. At the same time, the pointer will be live and can lead to UB if there is a mutable borrow after calling this method.

The safe way to implement this is to map the Ref's data reference to a `TokenAccount` reference, so the `Ref` stays alive while the reference is alive.

```suggestion
pub(crate) fn load_token_account(account: &AccountInfo) -> Result<Ref<TokenAccount>, ProgramError> {
    let account_data = account.try_borrow_data()?;
    if !valid_token_account_data(&account_data) {
        return Err(ProgramError::InvalidAccountData);
    }
    // SAFETY: We've validated the account data structure above
    Ok(Ref::map(account_info.try_borrow_data()?, |data| unsafe {
        &*(data.as_ptr() as *const TokenAccount)
    })
}
```

febo commented on p-ata/src/recover.rs:
```diff
@@ -0,0 +1,367 @@
+    let transfer_data = build_transfer_checked_data(amount_to_recover, nested_mint_decimals);
+
+    let transfer_metas = &[
+        AccountMeta {
+            pubkey: recover_accounts.nested_associated_token_account.key(),
+            is_writable: true,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: recover_accounts.nested_mint.key(),
+            is_writable: false,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: recover_accounts.destination_associated_token_account.key(),
+            is_writable: true,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: recover_accounts.owner_associated_token_account.key(),
+            is_writable: false,
+            is_signer: true,
+        },
```
Can use the constructors:
```suggestion
        AccountMeta::writable(
                recover_accounts.nested_associated_token_account.key()
        ),
        AccountMeta::readonly(
                recover_accounts.nested_mint.key()
        ),
        AccountMeta::writable(
                recover_accounts.destination_associated_token_account.key()
        ),
        AccountMeta::readonly_signer(
                recover_accounts.owner_associated_token_account.key()
        ),
```

febo commented on p-ata/src/recover.rs:
```diff
@@ -0,0 +1,367 @@
+        data: &CLOSE_ACCOUNT_DATA,
+    };
+
+    invoke_signed(
+        &ix_close,
+        &[
+            recover_accounts.nested_associated_token_account,
+            recover_accounts.wallet,
+            recover_accounts.owner_associated_token_account,
+        ],
+        &[pda_signer],
+    )?;
+    Ok(())
```
nit:
```suggestion
    invoke_signed(
        &ix_close,
        &[
            recover_accounts.nested_associated_token_account,
            recover_accounts.wallet,
            recover_accounts.owner_associated_token_account,
        ],
        &[pda_signer],
    )
```

febo commented on p-ata/src/recover.rs:
```diff
@@ -0,0 +1,367 @@
+    )?;
+
+    let close_metas = &[
+        AccountMeta {
+            pubkey: recover_accounts.nested_associated_token_account.key(),
+            is_writable: true,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: recover_accounts.wallet.key(),
+            is_writable: true,
+            is_signer: false,
+        },
+        AccountMeta {
+            pubkey: recover_accounts.owner_associated_token_account.key(),
+            is_writable: false,
+            is_signer: true,
+        },
```
```suggestion
        AccountMeta::writable(
            recover_accounts.nested_associated_token_account.key()
        ),
        AccountMeta::writable(
            recover_accounts.wallet.key()
        ),
        AccountMeta::readonly_signer(
            recover_accounts.owner_associated_token_account.key()
        ),
```

febo commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,974 @@
+        known_token_account_len,
+    )?;
+
+    match create_accounts.rent_sysvar {
```
The `rent_sysvar` seems to be used only to compute the minimal rent but at this point you already have the `space`. You could calculate the rent requirement here and replace the `match` statement with something like:
```rust
let minimum_balance = if let Some(rent_sysvar_info) = create_accounts.rent_sysvar {
        let rent = unsafe { Rent::from_account_info_unchecked(rent_sysvar_info)? };
        rent.minimum_balance(space)
} else {
        Rent::get()?.minimum_balance(space)
};
create_and_initialize_ata(...)
```

febo commented on p-ata/src/processor.rs:
```diff
@@ -0,0 +1,974 @@
+    while better_bump > expected_bump {
+        let maybe_better_address = derive_address::<3>(seeds, Some(better_bump), program_id);
+        if is_off_curve(&maybe_better_address) {
+            log!("Canonical address does not match provided address. Canonical bump is {}, with address {}.", better_bump, &maybe_better_address);
```
Not sure if you are expecting the `log!` to print the pubkey in base58 – as it is it will print the bytes.

rustopian committed:
move benches out to different branch

rustopian committed:
rm deps only used by benches

rustopian committed:
readme mention bencher branch

rustopian committed:
rm bench deps from feature

rustopian committed:
rm mollusk tests etc. from this pr

rustopian committed:
rm test deps

rustopian committed:
update from mollusk tests branch (#31)

* build(deps-dev): bump zx from 8.6.1 to 8.6.2 (#104)

Bumps [zx](https://github.com/google/zx) from 8.6.1 to 8.6.2.
- [Release notes](https://github.com/google/zx/releases)
- [Commits](https://github.com/google/zx/compare/8.6.1...8.6.2)

---
updated-dependencies:
- dependency-name: zx
  dependency-version: 8.6.2
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps-dev): bump zx from 8.6.2 to 8.7.0 (#105)

Bumps [zx](https://github.com/google/zx) from 8.6.2 to 8.7.0.
- [Release notes](https://github.com/google/zx/releases)
- [Commits](https://github.com/google/zx/compare/8.6.2...8.7.0)

---
updated-dependencies:
- dependency-name: zx
  dependency-version: 8.7.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps-dev): bump zx from 8.7.0 to 8.7.1 (#107)

Bumps [zx](https://github.com/google/zx) from 8.7.0 to 8.7.1.
- [Release notes](https://github.com/google/zx/releases)
- [Commits](https://github.com/google/zx/compare/8.7.0...8.7.1)

---
updated-dependencies:
- dependency-name: zx
  dependency-version: 8.7.1
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* deps: Update Solana v2.3.4 and rust toolchain (#106)

#### Problem

The v2.3 Solana crates are out, and the toolchain on this repo is a bit
old.

#### Summary of changes

Bump the Solana version, crates, and rust toolchain.

* build(deps): bump solana-program from 2.2.1 to 2.3.0 (#108)

* build(deps): bump solana-program from 2.2.1 to 2.3.0

Bumps [solana-program](https://github.com/anza-xyz/solana-sdk) from 2.2.1 to 2.3.0.
- [Release notes](https://github.com/anza-xyz/solana-sdk/releases)
- [Commits](https://github.com/anza-xyz/solana-sdk/compare/sdk@v2.2.1...sdk@v2.3.0)

---
updated-dependencies:
- dependency-name: solana-program
  dependency-version: 2.3.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Fixup deprecations

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jon C <me@jonc.dev>

* build(deps): bump solana-program-test from 2.3.4 to 2.3.5 (#109)

Bumps [solana-program-test](https://github.com/anza-xyz/agave) from 2.3.4 to 2.3.5.
- [Release notes](https://github.com/anza-xyz/agave/releases)
- [Changelog](https://github.com/anza-xyz/agave/blob/master/CHANGELOG.md)
- [Commits](https://github.com/anza-xyz/agave/compare/v2.3.4...v2.3.5)

---
updated-dependencies:
- dependency-name: solana-program-test
  dependency-version: 2.3.5
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps): bump solana-program-test from 2.3.5 to 2.3.6 (#111)

Bumps [solana-program-test](https://github.com/anza-xyz/agave) from 2.3.5 to 2.3.6.
- [Release notes](https://github.com/anza-xyz/agave/releases)
- [Changelog](https://github.com/anza-xyz/agave/blob/master/CHANGELOG.md)
- [Commits](https://github.com/anza-xyz/agave/compare/v2.3.5...v2.3.6)

---
updated-dependencies:
- dependency-name: solana-program-test
  dependency-version: 2.3.6
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps-dev): bump zx from 8.7.1 to 8.7.2 (#112)

Bumps [zx](https://github.com/google/zx) from 8.7.1 to 8.7.2.
- [Release notes](https://github.com/google/zx/releases)
- [Commits](https://github.com/google/zx/compare/8.7.1...8.7.2)

---
updated-dependencies:
- dependency-name: zx
  dependency-version: 8.7.2
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* interface: Add instruction definition, refactor some (#113)

* interface: Add instruction definition, refactor some

#### Problem

In order to publish the v3 SDK crates and have them usable in Agave, we
also need to have SPL crates using the v3 SDK crates. However, we have a
circular dependency between Agave and SPL which currently makes this
impossible.

The overall plan is to have Agave only use "interface" crates from SPL,
which have no dependencies on Agave crates.

You can see more info about the project at
https://github.com/orgs/anza-xyz/projects/27

ATA is already in a good position since it has a small "interface"-style
crate for Agave to use. However, it has two issues:

* the instruction definition is still in the program, and Agave needs
  that to deserialize instructions
* the name of the crate is "spl-associated-token-account-client", which
  is inconsistent with how we normally name these

#### Summary of changes

Move the instruction definition into the interface crate, gated with the
"borsh" feature for those who need it.

Rename the crate to "spl-associated-token-account-interface".

* Add required feature to borsh dep

* Run cargo fmt

* Bump version down to v1 for first release

* CI: Update git-cliff action to v4 (#114)

#### Problem

The v3 git-cliff action is very out of date, but still being used by this repo.

#### Summary of changes

Bump the git-cliff action version to v4.

* CI: Fix git-cliff args (#115)

#### Problem

The current yaml isn't actually recognized for the publish job.

#### Summary of changes

Remove some quotes. While I was at it, I also fixed the tag to `HEAD`
instead of `main`, which is useful in case we ever add backport
branches.

* build(deps-dev): bump typescript from 5.8.3 to 5.9.2 (#116)

Bumps [typescript](https://github.com/microsoft/TypeScript) from 5.8.3 to 5.9.2.
- [Release notes](https://github.com/microsoft/TypeScript/releases)
- [Changelog](https://github.com/microsoft/TypeScript/blob/main/azure-pipelines.release-publish.yml)
- [Commits](https://github.com/microsoft/TypeScript/compare/v5.8.3...v5.9.2)

---
updated-dependencies:
- dependency-name: typescript
  dependency-version: 5.9.2
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps-dev): bump zx from 8.7.2 to 8.8.0 (#117)

Bumps [zx](https://github.com/google/zx) from 8.7.2 to 8.8.0.
- [Release notes](https://github.com/google/zx/releases)
- [Commits](https://github.com/google/zx/compare/8.7.2...8.8.0)

---
updated-dependencies:
- dependency-name: zx
  dependency-version: 8.8.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* deps: Bump sdk to v3, decouple from program (#118)

#### Problem

The sdk v3 crates are out, but the ATA interface is still on v2.

#### Summary of changes

Bump the deps to v3, and make the program depend on the crates version
of the ATA interface to make the build possible.

* chore: Release

* build(deps): bump solana-program-test from 2.3.6 to 2.3.7 (#121)

Bumps [solana-program-test](https://github.com/anza-xyz/agave) from 2.3.6 to 2.3.7.
- [Release notes](https://github.com/anza-xyz/agave/releases)
- [Changelog](https://github.com/anza-xyz/agave/blob/master/CHANGELOG.md)
- [Commits](https://github.com/anza-xyz/agave/compare/v2.3.6...v2.3.7)

---
updated-dependencies:
- dependency-name: solana-program-test
  dependency-version: 2.3.7
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps): bump thiserror from 2.0.12 to 2.0.14 (#122)

Bumps [thiserror](https://github.com/dtolnay/thiserror) from 2.0.12 to 2.0.14.
- [Release notes](https://github.com/dtolnay/thiserror/releases)
- [Commits](https://github.com/dtolnay/thiserror/compare/2.0.12...2.0.14)

---
updated-dependencies:
- dependency-name: thiserror
  dependency-version: 2.0.14
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps): bump thiserror from 2.0.14 to 2.0.15 (#124)

Bumps [thiserror](https://github.com/dtolnay/thiserror) from 2.0.14 to 2.0.15.
- [Release notes](https://github.com/dtolnay/thiserror/releases)
- [Commits](https://github.com/dtolnay/thiserror/compare/2.0.14...2.0.15)

---
updated-dependencies:
- dependency-name: thiserror
  dependency-version: 2.0.15
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps): bump thiserror from 2.0.15 to 2.0.16 (#126)

Bumps [thiserror](https://github.com/dtolnay/thiserror) from 2.0.15 to 2.0.16.
- [Release notes](https://github.com/dtolnay/thiserror/releases)
- [Commits](https://github.com/dtolnay/thiserror/compare/2.0.15...2.0.16)

---
updated-dependencies:
- dependency-name: thiserror
  dependency-version: 2.0.16
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* build(deps-dev): bump zx from 8.8.0 to 8.8.1 (#127)

Bumps [zx](https://github.com/google/zx) from 8.8.0 to 8.8.1.
- [Release notes](https://github.com/google/zx/releases)
- [Commits](https://github.com/google/zx/compare/8.8.0...8.8.1)

---
updated-dependencies:
- dependency-name: zx
  dependency-version: 8.8.1
  dependency-type: direct:development
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

* checking in; branching out 3.0 update

* use 3.0 crates

* specify solana ver

* try cache bust

* try dedicated

* update mollusk

* passing mollusk tests

* error matching and other various test improvements

* rm old patches

* use 3.0 crates (#130)

* store token programs for now (until their repos use 3.0)

* fmt

* Use 3.0 crates (#29)

*remove CI hack

* sync and resolve conflicts (#30)

* fix merged file

* fmt

* specify tokio

* lockfile

* fmt, clippy

* rm unused

* process_and_merge_instruction helper

* ensure_system_account_exists helper for mollusk

* few more helpers reducing LoC

* a few more helpers

* lint

* rename vars to mollusk_result

* some small test fixes

* use mollusk check

* split up huge fn

* a little reuse

* CreateAtaInstructionType default

* rm async tokio

* test calc helpers

* move tests to harness

* rm some unused

* rm now-unused utils

* use harness for extended mint

* use breakout crates in tests

* minor clarifications and cleanups

* some naming updates and small utils cleanup

* rm redundant mint setup

* rm old comment

* rm solana-program-test dep

* rm dup

* rm intermediate harness

* rm unused fns

* use account builder in more places

* rm nested mod

* rm unneeded interface dev deps

* rm unrelated program import changes

* rm worthless wrappers

* rm specifiers

* rm old comment from comparison version

* rm submodules plan; program bins in fixtures

* more mollusk checks to ensure nested ata is empty and closed

* rm unnnecessary branching and specifiers; other small updates

AtaTestHarness capitalization, token_2022_interface/token_interface branching rm when unneeded, rm unnecessary specifiers, wrong-account helper takes Pubkey, deprecated instruction test adds rent sysvar account, setup_mollusk_with_programsnow points to programs/tests/fixtures

* move test harness to own crate

* cargo lock/toml updates

* rm unneeded branch: StateWithExtensionsOwned for both programs

* clearer name

* add ATA to spellcheck

* spellcheck

* merge create_ata_for_owner and create_nested_ata

* remove unnecessary path setters

* rm import

* accurate mint authority comment

* no sense having 0 lamports even when account expected to exist

* unified helpers for smaller token/token-2022 tests

* test-harness cargo.toml updates

* use mollusk mint constructors in some locations

* import lengths

* rm now unnecessary prog acct prefill

* rm hardcoded native loader and sys prog insertion

* update ctx_ensure_account_exists_with_lamports

* rm rent prefill

* use helper for token2022 too

* rm keypair

* ren_sysvar() no longer needed, woot

* fmt

* use interface

* rm unnecessary account insertions

* simplify some setup code

* simplify wallet_and_mint helper

* rm now unused

* rm stray keypair

* lint

* fmt

* comment

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Jon C <me@jonc.dev>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

rustopian committed:
update to 3.0 create-account-allow-prefund branches (#33)

rustopian committed:
Sync (#34)

rustopian committed:
Merge branch 'solana-program:main' into p-ata

rustopian committed:
rm submodules

rustopian committed:
Merge branch 'solana-program:main' into p-ata

grod220 commented:
All functionality in this PR is now represented on main in `pinocchio/`. Thanks @rustopian!
